lib/self-evict.js

```diff
@@ -182,65 +183,53 @@ SelfEvict.prototype._preEvict = function _preEvict() {
 
 SelfEvict.prototype._evict = function _evict() {
     var self = this;
+    var ringpop = self.ringpop;
+    var phase = this._transitionTo(PhaseNames.Evicting);
 
-    this._transitionTo(PhaseNames.Evicting);
+    ringpop.membership.setLocalStatus(Member.Status.faulty);
 
     var pingEnabled = this.ringpop.config.get('selfEvictionPingEnabled');
-
-    var restartGossip = false;
-    // Stop gossip when currently gossiping and ping is enabled
-    if (!this.ringpop.gossip.isStopped && pingEnabled) {
-        this.ringpop.gossip.stop();
-        restartGossip = true;
-    }
-
     if (pingEnabled) {
-        waitUntilNotPinging();
-
-        function waitUntilNotPinging() {
-            if (self.ringpop.gossip.isPinging) {
-                //Wait until isPinging is false...
-                setTimeout(waitUntilNotPinging, 25);
-            } else {
-                self.ringpop.membership.setLocalStatus(Member.Status.faulty);
-                gossipEviction();
-            }
-        }
-
-        function gossipEviction() {
-            var selfEvictionMaxPingRatio = self.ringpop.config.get('selfEvictionMaxPingRatio');
-            var membership = self.ringpop.membership;
-
-            var numberOfPingableMembers = _.filter(membership.members, membership.isPingable.bind(membership)).length;
-            var maxNumberOfTicks = Math.ceil(numberOfPingableMembers * selfEvictionMaxPingRatio);
-
-            // Never ping more than the maximum piggy back counter.
-            var numberOfPings = Math.min(
-                self.ringpop.dissemination.maxPiggybackCount,
-                maxNumberOfTicks
-            );
+        var selfEvictionMaxPingRatio = ringpop.config.get('selfEvictionMaxPingRatio');
+        var membership = ringpop.membership;
 
-            tick();
+        var pingableMembers = _.filter(membership.members, membership.isPingable.bind(membership));
+        var numberOfPingableMembers = pingableMembers.length;
+        var maxNumberOfTicks = Math.ceil(numberOfPingableMembers * selfEvictionMaxPingRatio);
 
-            function tick() {
-                if (numberOfPings > 0) {
-                    numberOfPings--;
-                    self.ringpop.gossip.tick(tick);
-                } else {
-                    done();
-                }
+        // Never ping more than the maximum piggy back counter.
+        var numberOfPings = Math.min(
```

Comment:
lets also add `numberOfPingableMembers` to the min. There are cases where `numberOfPings` might exceed the number of pingable members.

First the `selfEvictionMaxPingRatio` must be bigger then 1
Secondly the `maxPiggybackCount` needs to exceed the number of members. This is true on cluster sizes less then 30 I believe.